master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql

```diff
@@ -0,0 +1,93 @@
+CREATE UNIQUE INDEX ix_metrics_name_experiment_id_unique ON exp_metrics_name(experiment_id);
+
+INSERT INTO public.exp_metrics_name (project_id, experiment_id, tname, vname) (
+    WITH training_metrics_names AS (
+        SELECT array_to_json(array_agg(DISTINCT names)) AS name, e.id AS experiment_id
+        FROM trials t, experiments e, steps s,
+            LATERAL jsonb_object_keys(s.metrics->'avg_metrics') AS names
+        WHERE t.id=s.trial_id AND e.id = t.experiment_id 
+        GROUP BY e.id),
+    validation_metrics_names AS (
+        SELECT array_to_json(array_agg(DISTINCT names)) AS name, e.id AS experiment_id
+        FROM trials t, experiments e, validations v,
+            LATERAL jsonb_object_keys(v.metrics->'validation_metrics') AS names
+        WHERE t.id=v.trial_id AND e.id = t.experiment_id 
+        GROUP BY e.id)
+    SELECT   
+        e.project_id AS project_id, 
+        e.id AS experiment_id, 
+        COALESCE(training_metrics_names.name, '[]'::json) AS tname,
+        COALESCE(validation_metrics_names.name, '[]'::json) AS vname 
+    FROM experiments e LEFT JOIN validation_metrics_names ON e.id = validation_metrics_names.experiment_id, training_metrics_names 
+    WHERE 
+        training_metrics_names.experiment_id = e.id
+);
+
+CREATE OR REPLACE FUNCTION autoupdate_exp_training_metrics_name() RETURNS trigger AS $$
+BEGIN
+    INSERT INTO exp_metrics_name (project_id, experiment_id, tname) (
+        SELECT e.project_id, e.id, array_to_json(array_agg(DISTINCT names)) AS tname
+        FROM experiments e, trials t, raw_steps s, LATERAL jsonb_object_keys(s.metrics->'avg_metrics') AS names
+        WHERE  t.experiment_id = (SELECT experiment_id FROM trials WHERE id = NEW.trial_id) AND t.id = s.trial_id AND e.id=t.experiment_id GROUP BY (e.project_id, e.id)
+    )  ON CONFLICT(experiment_id) DO UPDATE SET tname = EXCLUDED.tname;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER autoupdate_exp_training_metrics_name
+AFTER INSERT ON raw_steps
+FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_training_metrics_name();
+
+CREATE OR REPLACE FUNCTION autoupdate_exp_validation_metrics_name() RETURNS trigger AS $$
+BEGIN
+    INSERT INTO exp_metrics_name (project_id, experiment_id, vname) (
+        SELECT e.project_id, e.id, array_to_json(array_agg(DISTINCT names)) AS vname
+        FROM experiments e, trials t, raw_validations v, LATERAL jsonb_object_keys(v.metrics->'validation_metrics') AS names
+        WHERE  t.experiment_id = (SELECT experiment_id FROM trials WHERE id = NEW.trial_id) AND t.id = v.trial_id AND e.id=t.experiment_id GROUP BY (e.project_id, e.id)
+    )  ON CONFLICT(experiment_id) DO UPDATE SET vname = EXCLUDED.vname;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER autoupdate_exp_validation_metrics_name
+AFTER INSERT ON raw_validations
+FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_validation_metrics_name();
+
```

Comment:
I thought we were not going to include metrics here until summary metrics are done: how would these metrics columns be useful if we can't do any filtering/sorts using them? cc @hkang1 

There's a work-in-progress proposal for summary metrics structure [in a "Summary Metrics" section here](https://hpe.sharepoint.com/:w:/r/teams/detai/Shared%20Documents/Engineering/Backend/Unified%20metrics/ERD-%20Unified%20Metrics.docx?d=w6ff07466cd42461593056a8aae14a7e6&csf=1&web=1&e=hfsoed). it'll effectively supercede this table.

Reply:
@ioga does summary metrics include validation metrics? I thought the columns should still at least return something like:
```
metrics: [
  'validation.accuracy',
  'validation.loss',
],
```
and it would exclude things like `training.accuracy` and `training.loss` until summary metrics work is done.

Reply:
today there's only one summary metric: best searcher metric, which is a validation metric. other validation metrics aren't included.

based on my chat with Guangqing, these columns will be used in place of existing `MetricNames` to get the column names for *time series* requests. if these column names are intended for filtering, we need to do this differently.

Reply:
I believe filtering functionality will come after the unified metrics work, right now we just need these metrics name in place of `MetricNames` call